### ISSUE TITLE
Cut note(s) loss protection

### ISSUE THREAD

**Issue by VeH-c** [orgzly-revived/orgzly-android-revived#204]
When you cut a note it disappears, then when you go to paste it, if you Copy or Cut instead, the previously Cut note is lost. 
This is an old issue that I think should be fixed/mitigated, as losing data trying to hit the right icon is a very frustrating experience.
https://github.com/orgzly/orgzly-android/issues/662
I do not know what the fix or mitigation should be, but the above issues does have some good suggestions as a place to start,

Related
https://github.com/orgzly/orgzly-android/issues/659

**Comment by amberin** [comment]
I think this should be considered a bug, rather than a missing feature. The most expected behavior is for "cut" actions to not be destructive in this way.

**Comment by ShuoShen** [comment]
@amberin What is a typical use for cutting (or copying and pasting a note? I could only imagine moving the tree/sub-tree around. If that's the main use, maybe we can improve the move-tree tree experience. I can't really think of any other use case TBH.

**Comment by amberin** [comment]
> [@amberin](https://github.com/amberin) What is a typical use for cutting (or copying and pasting a note? I could only imagine moving the tree/sub-tree around. If that's the main use, maybe we can improve the move-tree tree experience, and remove the cut option (or at least hide copy/cut when there's one already in the clipboard). I can't really think of any other use case TBH.

By move-tree, do you mean using the arrows? Cutting allows moving in different ways, most notably across notebooks.

**Comment by KaKi87** [comment]
The issue is with cutting, not copying.

**Comment by ShuoShen** [comment]
> The issue is with cutting, not copying.

@KaKi87, yes that's my understanding as well. Can you help me confirm if this is the sequence that gets you into a data loss issue? 

cut a tree -> accidentally hit copy (or less likely cut) -> lose previous cut tree

The point is that the app is doing things correctly. But it's confusing, isn't it? If that's the case, then we shouldn't change the app behavior, but design the app so that it's harder for people to click the wrong button, or provide a way for them to undo the cut.

**Comment by amberin** [comment]
I guess it depends on how you define "bug", but it's a very poor experience, and not at all how you'd expect a "cut" feature to behave. You wouldn't expect losing data. So for practical purposes, it's a bug that needs fixing.

**Comment by ShuoShen** [comment]
> I guess it depends on how you define "bug", but it's a very poor experience, and not at all how you'd expect a "cut" feature to behave. You wouldn't expect losing data; you'd expect your previously cut content to get put back where it was. So for practical purposes, it's a bug that needs fixing.

Fair enough. Thanks for confirmation. 

We need a proper UX behavior for this interaction then. 

I think there needs to be some sort of undo button, otherwise we would still run into the problem even with better UI design. 

And on top of that, the action buttons would work better with captions. This is a problem with not just the copy/paste bar, but most of the action bars in the app (e.g. the one at the bottom in the previous screenshot). It would be nice to have a designer to help us with the design.

**Comment by amberin** [comment]
I believe preferring icons over text is a very common paradigm in mobile apps. But there should always be an associated text label for accessibility reasons. If you hold the copy or cut ivon, you see a caption.

**Comment by ShuoShen** [comment]
> I believe preferring icons over text is a very common paradigm in mobile apps. But there should always be an associated text label for accessibility reasons. If you hold e.g. the copy or cut button, you see a caption.

Really? I was having a different impression. Looking at YouTube, Reddit and GitHub, and they all have texts under buttons at the bottom nav bar. Google Photo has text for edit operations also at the bottom. However I dont see any text when the buttons are at the top bar.

**Comment by amberin** [comment]
@ShuoShen You may very well be right - I'm no expert, I'm sure there are many others here (such as yourself) who may be more knowledgeable. But that's just been my general impression.

**Comment by KaKi87** [comment]
Hi,
Any news on this ?
Thanks

**Comment by KaKi87** [comment]
@amberin 🙏

**Comment by KaKi87** [comment]
@emilymclean 🙏

**Comment by emilymclean** [comment]
@KaKi87 I am not a maintainer of this project and this is not a particularly motivating way to ask me to do something for you. In my opinion this could be trivially fixed by only removing a cut note only after the user pastes. To me, admittedly being unfamiliar with this part of the codebase, this actually looks like a pretty simple first issue that I would motivate you to try and pick up yourself if @amberin is ok with my solution. 

If you removed the delete statement from [this file](https://github.com/orgzly-revived/orgzly-android-revived/blob/31856ce8178a7b05a63e5e20aaf74e92bb81e0ea/app/src/main/java/com/orgzly/android/usecase/NoteCut.kt#L12), added a property to [NotesClipboard.Entry](https://github.com/orgzly-revived/orgzly-android-revived/blob/31856ce8178a7b05a63e5e20aaf74e92bb81e0ea/app/src/main/java/com/orgzly/android/db/NotesClipboard.kt#L16) to indicate if the clipboard operation is a cut, then added back the delete code to [here](https://github.com/orgzly-revived/orgzly-android-revived/blob/31856ce8178a7b05a63e5e20aaf74e92bb81e0ea/app/src/main/java/com/orgzly/android/usecase/NotePaste.kt#L11) guarded by "if entry is a cut" (plus updated all the tests) you'd be done. 

The real work will almost inevitably be checking whether this change has unintended side effects. Like I'm not sure why NotesClipboard has the ability to hold multiple entries, but maybe that is important because of something I'm unaware of.

**Comment by KaKi87** [comment]
> I am not a maintainer of this project and this is not a particularly motivating way to ask me to do something for you.

I'm sorry, as it happens [you're the only significant contributor to this project](https://github.com/user-attachments/assets/26a11831-2e04-4433-a12e-6acec17c232a), and I'm desperate, I reported this issue over a year and a half ago, OP did over 2 years and a half ago, and someone on the original project reported this issue over 6 and a half years ago, and to this day, I keep permanently loosing precious data a few times a month because of this issue 😭 

> this actually looks like a pretty simple first issue that I would motivate you to try and pick up yourself

As I have no experience whatsoever in Android app development, in using compiled languages, and have no machine powerful with enough RAM and storage to do that, the only thing I could reasonably do is have a Cursor AI cloud agent file a PR, which I wouldn't be able to review myself 😕 

> The real work will almost inevitably be checking whether this change has unintended side effects.

However I'd be fine with long-term testing a nightly build containing the fix, and reporting any issue I'd find along the way.

**Comment by emilymclean** [comment]
@KaKi87 I think looking at a code chart significantly downplays the efforts of other contributors here, particularly amberin. Managing an open source project long term is no small task and I don't want people overlooking their excellent work. I, on the other hand, am simply fixing or improving aspects of the app I personally find problematic. 

Unfortunately I am busy starting a new job right now, so I don't have time to fix your issue. However, I can see from your profile that you're already skilled in at least some programming and the core work of this change is about 5 lines of code - the contents of which I've already described. As a gesture of good faith, if you are willing to get the ball rolling here and make a draft PR with these changes I would be more than happy to help you get it across the line.

**Comment by KaKi87** [comment]
> you're already skilled in at least some programming

Sure, I do Node JS and browser JS, that's it.

> the core work of this change is about 5 lines of code - the contents of which I've already described. As a gesture of good faith, if you are willing to get the ball rolling here and make a draft PR with these changes I would be more than happy to help you get it across the line

Done : #1027
